File: experiments.py
"""
==============================================================================
Exeriment Set
------------------------------------------------------------------------------


Analyse runtime characterists : CPP vs Python. 


: 03 August 2023
: Zach Wolpe
: zach.wolpe@medibio.com.au
==============================================================================
"""

from DSA.DataStructures.DoublyLinkedLists.DoublyLinkedList import *
from DSA.DataStructures.StacksQueues.StacksAndQueues import *
from DSA.DataStructures.LinkedLists.LinkedList import *
from DSA.DataStructures.BinarySearchTree.bst import *
from DSA.DataStructures.HashTable.HashTable import *
from DSA.DataStructures.Graphs.graphs import *

from DSA.Algorithms.python.selection_sort import *
from DSA.Algorithms.python.insertion_sort import *
from DSA.Algorithms.python.tree_traverse import *
from DSA.Algorithms.python.bubble_sort import *
from DSA.Algorithms.python.merge_sort import *
from DSA.Algorithms.python.quick_sort import *


from Profiler.pyModules.profiler        import CPUProfiler, MemoryProfiler
from Profiler.pyModules.dependencies    import *

from helpers import *
import pandas as pd
import numpy as np
import random
import psutil
import time
import sys
import os


def sorting_algorithms(array, *args, **kwargs):

    # standard sorting algorithms
    selection_sort(array)
    insertion_sort(array)
    bubble_sort(array)
    merge_sort(array)
    # quickSort.quick_sort(array)

    # build tree
    tree = TreeTraversal()
    [tree.insert(i) for i in array]

    # # tree traverse
    tree.BFS()
    tree.DFS('preorder')
    tree.DFS('inorder')
    tree.DFS('postorder')
    return True


def datastructures(array, *args, **kwargs):
    
    # # graphs
    my_graph = Graph()
    print('_dir_: ', dir(my_graph))
    for _ in range(len(array)):
        my_graph.addVertex('A')
        my_graph.addEdge('A','B')
        my_graph.removeVertex('A')

    # Binary Search Tree
    my_tree = BinarySearchTree()
    [my_tree.insert(a) for a in array]

    # Hash Table
    my_hash = HashTable()
    [my_hash.set_item(str(a), a) for a in array]

    # Queue
    my_queue = Queue(33)
    [my_queue.enqueue(a) for a in array]
    
    return True


@CPUProfiler.profile 
@MemoryProfiler.profile
def core_algorithms(array, *args, **kwargs):
    # core algorithms
    datastructures(array)
    sorting_algorithms(array)
    return True


from datetime import datetime, timedelta


class basic_timer:
    def __init__(self, *args, **kwargs) -> None:
        self.start_time = time.time()
        self.end_time   = None

    def __enter__(self, *args, **kwargs):
        return self

    def __exit__(self, *args, **kwargs):
        self.end_time = time.time()
        # convert to seconds
        self.time_delta = self.end_time - self.start_time
        self.delta      = timedelta(seconds=self.time_delta)

        print('time taken: ', self.delta)
        return self
    
    

def write_to_file(_dict, path):
    with open(f'{path}/basetimes.pkl', 'wb') as f:
        pickle.dump(_dict, f)




if __name__ == '__main__':
    clear_directory()

    # Run experiments
    store_loc  = 'results'
    base_times = {}

    for _, size in enumerate([10, 100, 1000, 10000, 100000]):
        bt = basic_timer()
        with bt:
            print('launching size=', size)
            array = np.random.sample(size)*100 // 1
            result, mem_profile = core_algorithms(array, sav_loc=f'{store_loc}/n_combinations-{size}')
            print('\nresult:    ', result)
            print('mem_profile: ', mem_profile)
            print(array)
        base_times[size] = bt.delta

    print('base_times: ', base_times)
    write_to_file(base_times,store_loc)
        

File: experiments.cpp
/*
==============================================================================
Exeriment Set
------------------------------------------------------------------------------


Analyse runtime characterists : CPP vs Python. 


: 03 August 2023
: Zach Wolpe
: zach.wolpe@medibio.com.au
==============================================================================
*/


#include <iostream>
#include <string>
#include <vector>
#include <chrono>
#include <array>
#include <fstream>

#include <chrono>
#include <fstream>
#include <map>
#include <cstdlib> / for exit function

#include "DSA/Algorithms/cpp/bubble_sort.cpp"
#include "DSA/Algorithms/cpp/selection_sort.cpp"
#include "DSA/Algorithms/cpp/quick_sort.cpp"
#include "DSA/Algorithms/cpp/merge_sort.cpp"
#include "DSA/Algorithms/cpp/insertion_sort.cpp"
#include "DSA/Algorithms/cpp/tree_traverse.cpp"
#include "Profiler/cppModules/profiler.cpp"



void tree_sorter_inorder(int array[], int size) {
    BinarySearchTree* BST = new BinarySearchTree();
    for (int i=0; i<size; i++) {
        BST->insert(array[i]);
    };
    BST->DFS("inorder");
};

void tree_sorter_postorder(int array[], int size) {
    BinarySearchTree* BST = new BinarySearchTree();
    for (int i=0; i<size; i++) {
        BST->insert(array[i]);
    };
    BST->DFS("postorder");
};

void tree_sorter_preorder(int array[], int size) {
    BinarySearchTree* BST = new BinarySearchTree();
    for (int i=0; i<size; i++) {
        BST->insert(array[i]);
    };
    BST->DFS("preorder");
};

void merge_sorter(int array[], int size) {
    int leftIndex   = 0;
    int rightIndex  = size - 1;
    mergeSort(array, leftIndex, rightIndex);
};

void quick_sorter(int array[], int size) {
    int leftIndex   = 0;
    int rightIndex  = size - 1;
    quickSort(array, leftIndex, rightIndex);
};

void bubbleSort(int array[], int size);
void selectionSort(int array[], int size);
void insertionSort(int array[], int size);


bool sorting_algorithms(int arr_size, void (*algorithm)(int array[], int size)) {
    PROFILE_FUNCTION();
    int myArray[arr_size];
    for(int i=0;i<arr_size;i++)
        myArray[i]=rand()%100;
    int size = sizeof(myArray) / sizeof(myArray[0]);
    algorithm(myArray, size);
    // PROFILE_FUNCTION();
    return true;
    // Instrumentor::Get().EndSession();
    // For adhoc debugging.
    // std::cout << " myArray : " << myArray << ": ";
    // for(int i=0;i<arr_size;i++)
    //     std::cout << myArray[i] << " ";
    // std::cout << std::endl;
};





class basic_timer {
    public:
        std::chrono::time_point<std::chrono::high_resolution_clock> start_time;
        std::chrono::time_point<std::chrono::high_resolution_clock> end_time;
        bool running;

        void start() {
            if (!running) {
                start_time = std::chrono::high_resolution_clock::now();
                end_time   = start_time;
                running    = true;
            }
        };

        void stop() {
            if (running) {
                end_time = std::chrono::high_resolution_clock::now();
                running  = false;
            }
        };

        void restart() {
            start_time = std::chrono::high_resolution_clock::now();
            running = false;
        }

        double elapsed() {
            if (running) {
                return std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - start_time).count();
            } else {
                return std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();
            }
        };
};





class write_to_file {

public:
    std::ofstream outdata;
    void open_file(std::string loc = "runtime-cpp.txt") {
        this->outdata.open(loc);
        if (!this->outdata) {
            std::cerr << "Error: file could not be opened" << std::endl;
            exit(1);
        }
    };
    void close_file() {
        this->outdata.close();
    };
    void write_data(int key, int value) {
        this->outdata << "{" << key << ", " << value << "}, " << std::endl;
    };
};



void runtime() {
    PROFILE_FUNCTION();
// write to file
    write_to_file file;
    file.open_file();
    basic_timer timer;
    // run experiments
    for (int x=10; x<=100000; x*=10) {
        timer.start();
        std::cout << "x : " << x << std::endl;
        sorting_algorithms(x, &bubbleSort);
        sorting_algorithms(x, &selectionSort);
        sorting_algorithms(x, &insertionSort);
        sorting_algorithms(x, &merge_sorter);
        sorting_algorithms(x, &quick_sorter);
        sorting_algorithms(x, &tree_sorter_preorder);
        sorting_algorithms(x, &tree_sorter_inorder);
        sorting_algorithms(x, &tree_sorter_postorder);
        timer.stop();
        std::cout << "x : " << x << std::endl;
        std::cout << "time : " << timer.elapsed() << std::endl;
        file.write_data(x, timer.elapsed());
    }
};


void runtime_profiler() {
    PROFILE_FUNCTION();
    std::cout << ">> Running programWrapper..." << std::endl;
    Instrumentor::Get().LaunchSession("myprofiler", "profiler-v2x.json"); 
    runtime();
    Instrumentor::Get().EndSession();

};


int main() {
    runtime_profiler();
    return 0;
};


